File: src/types/shape.ts
export type ShapeType =
  | "rectangle"
  | "triangle"
  | "lshape"
  | "circle"
  | "trapezoid"

export type RotationAngle = 0 | 90 | 180 | 270

export interface ShapeField {
  key: string
  label: string
  unit?: string
  min?: number
  max?: number
  step?: number
}

export interface ShapeDefinition {
  type: ShapeType
  label: string
  defaultParameters: Record<string, number>
  fields: readonly ShapeField[]
}

export interface ShapeConfig {
  type: ShapeType
  parameters: Record<string, number>
  rotation: RotationAngle
  flipX: boolean
  flipY: boolean
}

File: src-tauri/src/models/shape.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ShapeConfig {
    #[serde(rename = "type")]
    pub shape_type: ShapeType,

    pub parameters: HashMap<String, f64>,

    pub rotation: u16,

    #[serde(rename = "flipX")]
    pub flip_x: bool,

    #[serde(rename = "flipY")]
    pub flip_y: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum ShapeType {
    Rectangle,
    Circle,
    Triangle,
    Lshape,
    Trapezoid,
}